File: c++/0063.cpp
// Link: https://leetcode.com/problems/unique-paths-ii/
// Runtime: 0 ms
// Memory: 7.5 MB

class Solution {
public:
  int uniquePathsWithObstacles(std::vector<std::vector<int>> &obstacleGrid) {
    int m = obstacleGrid.size(), n = obstacleGrid[0].size();
    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++)
        if (obstacleGrid[i][j] == 1)
          obstacleGrid[i][j] = -1;

    if ((obstacleGrid[m - 1][n - 1] == 0) && (obstacleGrid[0][0] == 0)) {
      obstacleGrid[0][0] = 1;

      for (int i = 0; i < m; i++)
        for (int j = 0; j < n; j++)
          if (obstacleGrid[i][j] != -1) {
            if ((i - 1 > -1) && (obstacleGrid[i - 1][j] != -1))
              obstacleGrid[i][j] += obstacleGrid[i - 1][j];
            if ((j - 1 > -1) && (obstacleGrid[i][j - 1] != -1))
              obstacleGrid[i][j] += obstacleGrid[i][j - 1];
          }
    } else
      obstacleGrid[m - 1][n - 1] = 0;

    return obstacleGrid[m - 1][n - 1];
  }
};

File: python/0063.py
# Link: https://leetcode.com/problems/unique-paths-ii/description/
# Runtime: 47 ms
# Memory: 16.45 MB


class Solution:
    def uniquePathsWithObstacles(self, obstacleGrid: List[List[int]]) -> int:
        paths = [([0] * len(obstacleGrid[0])) for _ in range(len(obstacleGrid))]
        if obstacleGrid[len(obstacleGrid) - 1][len(obstacleGrid[0]) - 1] == 0:
            paths[len(obstacleGrid) - 1][len(obstacleGrid[0]) - 1] = 1

        for i in range(len(paths) - 1, -1, -1):
            for j in range(len(paths[0]) - 1, -1, -1):
                if (obstacleGrid[i][j] == 0) and (i != len(paths) - 1):
                    paths[i][j] += paths[i + 1][j]
                if (obstacleGrid[i][j] == 0) and (j != len(paths[0]) - 1):
                    paths[i][j] += paths[i][j + 1]

        return paths[0][0]
